### PR TITLE
__arglist cannot have an argument passed by 'in' or 'out'

### LINKED ISSUE #22455: `in` and `out` arguments should be disallowed in `__arglist`
varargs should disallow  `out` and `in` arguments. There is no way the calee can express its part of the contract.

We currently allow both and `out` was allowed in old compiler versions too. We should disallow both. 

Arguably it would be a breaking change for `out`, but it is very unlikely it would affect real code. 

Example:

```C#

    class Program
    {
        static void Main(string[] args)
        {
            int unassigned;
            int doNotChange = 42;

            Test(__arglist(out unassigned, in doNotChange));  // <--  out?   in?  really?

            System.Console.WriteLine(unassigned);
            System.Console.WriteLine(doNotChange);
        }

        static void Test(__arglist)
        {
            var args = new ArgIterator(__arglist);

            // leave unassigned unassigned
            ref var unassigned = ref __refvalue(args.GetNextArg(), int);

            // assign junk to doNotChange
            ref var doNotChange = ref __refvalue(args.GetNextArg(), int);
            doNotChange = unassigned;
        }
    }
```

### PR DESCRIPTION
**Pull request by OmarTawfik**
Fixes #22455

cc @VSadov @dotnet/roslyn-compiler

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs [hunk 1]
```diff
@@ -199,6 +199,19 @@ private BoundExpression BindArgListOperator(InvocationExpressionSyntax node, Dia
...
                     Error(diagnostics, ErrorCode.ERR_CantUseVoidInArglist, argument.Syntax);
                 }
+
+                switch (analyzedArguments.RefKind(i))
+                {
+                    case RefKind.None:
+                    case RefKind.Ref:
+                        break;
+                    case RefKind.In:
+                    case RefKind.Out:
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs]: Does the `out` change need to be recorded as a breaking change?

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs [hunk 2]
```diff
@@ -199,6 +199,19 @@ private BoundExpression BindArgListOperator(InvocationExpressionSyntax node, Dia
...
+                        Error(diagnostics, ErrorCode.ERR_CantUseInOrOutInArglist, argument.Syntax);
+                        break;
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(analyzedArguments.RefKind(i));
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs]: Consider extracting a local for `analyzedArguments.RefKind(i)`.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs]: >throw ExceptionUtilities.UnexpectedValue(analyzedArguments.RefKind(i)); [](start = 24, length = 71)

I think it is better to not throw here, but instead check for kinds that we support. We support ```None``` and ```Ref```, for everything else it is safe to report an error and I believe it is better than throwing.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2).

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs (hunk 2 above)
Comment (OmarTawfik) [reply src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs]: What do you suggest to put in the diagnostic message instead? the notion of "refkind" is not public or intuitive, and the current one only mentions `in` and `out`.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs]: > What do you suggest to put in the diagnostic message instead? the notion of "refkind" is not public or intuitive, and the current one only mentions in and out.

I suggest to go with the message you have. It covers what we have today and, once we add a new kind, we either adjust the condition of the message.

**Comment by OmarTawfik** [comment]
@cston @AlekseyTs comments addressed.

**Pull request merged**